Add getting started section (#9)
OSOceanAcoustics/echoshader: gsoc_contributors_guide.md

## Getting Started

> 

Reviewer: I have a general suggestion here that it would be good to limit the line length so that the text is easier to read, especially on a small screen the wrap-around can be disruptive. See the recommendation in [black](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length). For echopype we have it [configured to 100](https://github.com/OSOceanAcoustics/echopype/blob/5cf1153b3d292ce819276ed12c86816c694b75ec/setup.cfg#L21-L25) but I would recommend maybe 90.

Author: @leewujung how exactly do I enable the limiting of the line length on Github for a markdown file? The links you provided seem to be Python specific.

Reviewer: I pointed to those just as references for why it's good to do so, but I was thinking to just break it manually here would be fine, since in this repo the main work will still be code and not text.

But, I did a quick google search when seeing your question. I think we can use tools such as [markdownlint](`https://github.com/markdownlint/markdownlint`) as part of pre-commit so that this is automated also. 

I'll make an issue for creating pre-commit and config for these. We don't have code here just yet for black or flake8 to lint, but maybe can set up markdownlint to get the mechanism in place. Interested? ;)

Author: I've never set these up on my own, but I can give it a try. Go ahead and assign me to the issue you create.

The sonar data we will be working with comes from several instruments and are stored in packets of data called datagrams. 

> However, the structure of these datagrams are difficult to work with and do not allow for efficient processing 

Reviewer: ```suggestion
However, these datagrams are stored in binary format, which can be difficult to work with and do not allow for efficient processing 
```

> The sonar data we will be working with comes from several instruments and are stored in packets of data called datagrams. 
> However, the structure of these datagrams are difficult to work with and do not allow for efficient processing 
> of the data. To put these datagrams in a more user friendly format, we utilize echopype. Echopype converts the data into a EchoData
> object. This EchoData object is beneficial as it structures the data into an interoperable netCDF sonar data format and 
> enables processing that is scalable. The resulting calibration of this EchoData object will be the ocean sonar data you 
> will be working with. Before diving into the project, we suggest you review a couple of preliminary 
> items. We outline these items below as well as provide some additional helpful resources and initial steps to get 
> started. 

Reviewer: ```suggestion
The sonar data we will be working with can come from several different instruments and are stored in different binary formats specific to these instruments. 
These binary data is difficult to work with directly and does not allow for efficient processing.
We use [echopype](https://github.com/OSOceanAcoustics/echopype) to convert the raw data into a more user friendly format following an interoperable netCDF data model, and serialize the data into netCDF or Zarr. 
These standardized raw data is then calibrated into Sv (volume backscattering strength -- how strong the echo return is in a volume of water) dataset you will be working with.

Before diving into the project, we suggest that you review the items below. 
We also provide some additional helpful resources and initial steps to get you started. 
```

### Storage format

We use two formats to store the data: 

> * netcdf files - the current defacto file for working with array-oriented scientific data from climate research.

Reviewer: ```suggestion
* [netCDF files](https://www.unidata.ucar.edu/software/netcdf/) - the current defacto file for working with array-oriented scientific data from climate research.
```

Although it is not necessary to understand the netCDF library in its entirety, Unidata does provide a well documented 

> [netcdf](https://unidata.github.io/netcdf4-python/) python interface. This documentation describes useful aspects of 

Reviewer: ```suggestion
[netcdf python interface](https://unidata.github.io/netcdf4-python/). This documentation describes useful aspects of 
```

> how netcdf defines common terms such as groups, dimensions, variables, and attributes.  

Reviewer: ```suggestion
how netCDF defines common terms such as groups, dimensions, variables, and attributes.  
```

* [Zarr](https://zarr.readthedocs.io/en/stable/) - a format for the storage of chunked, compressed, N-dimensional 

> arrays. Zarr has similar characteristics to netcdf, but has the added benefit of being a cloud-native data format. For 

Reviewer: ```suggestion
arrays. Zarr has similar characteristics to netCDF, but has the added benefit of being a cloud-native data format. For 
```

this reason, Zarr is ideal for storing large data sets in the cloud. Zarr provides a great overview of its [storage specifications](https://zarr.readthedocs.io/en/stable/spec/v2.html#hierarchies)
that may be useful to read.    

### Data Structures

Netcdf and Zarr formats can be easily read with the [xarray](https://xarray.pydata.org/en/stable/index.html) library in 
Python. Additionally, xarray enables efficient computation of our data, which is labelled and multi-dimensional. 
A fantastic [xarray tutorial](https://xarray-contrib.github.io/xarray-tutorial/) has been put 
together that describes the fundamentals of xarray. Be sure to become familiar with both DataArrays and Datasets as they
are heavily used.     

### Ocean Sonar Data

> For this project, you will be initially working with the output of the [compute_Sv](https://echopype.readthedocs.io/en/stable/api/echopype.calibrate.compute_Sv.html#echopype.calibrate.compute_Sv) 
> function. This is a function in echopype that computes the volume backscattering strength (Sv) of the provided EchoData
> object. This function returns an xarray Dataset that has several variables that are necessary for the visualization 
> of ocean sonar data. The Dataset has the dimensions and coordinates: 

Reviewer: ```suggestion
For this project, you will be initially working with output of the [compute_Sv](https://echopype.readthedocs.io/en/stable/api/echopype.calibrate.compute_Sv.html#echopype.calibrate.compute_Sv) 
function. This is a function in echopype that computes the volume backscattering strength (Sv, basically how strong the echo return is from a volume of water) from the raw data. This function returns an xarray Dataset that has several variables that are necessary for the visualization 
of ocean sonar data. The Dataset has the dimensions and coordinates: 
```

> function. This is a function in echopype that computes the volume backscattering strength (Sv) of the provided EchoData
> object. This function returns an xarray Dataset that has several variables that are necessary for the visualization 
> of ocean sonar data. The Dataset has the dimensions and coordinates: 

Reviewer: ```suggestion
function. This is a function in echopype that computes the volume backscattering strength (Sv) from the raw data. Sv is basically how strong the echo return is from a volume of water. This function returns an xarray Dataset that has several variables that are necessary for the visualization 
of ocean sonar data. The Dataset has the dimensions and coordinates: 
```

* range_bin - sample index along a range

The data variables of the Dataset are as follows, where items in parenthesis are the dimensions of the data variables: 

* Sv (frequency, ping_time, range_bin) - volume backscattering strength measured from the echo

* range (frequency, ping_time, range_bin) - the measured range of an echo in meters 

* temperature - the temperature measurement of the water collected by the echosounder, with unit degree Celsius

* salinity - the salinity measurement of the water collected by the echosounder, with unit part per thousand (PSU)

* pressure - the pressure measurement of the water collected by the echosounder, with unit dbars

* sound_speed (frequency, ping_time) - sound speed (in units m/s) for the provided temperature, salinity, and pressure 

* sound_absorption (frequency, ping_time) - sea water absorption (in units dB/m) for each frequency and ping time, this 
value is based on the temperature, salinity, pressure, and sound_speed

* sa_correction (frequency) - the sa correction for each frequency

* gain_correction - (frequency) - the gain correction for each frequency

* equivalent_beam_angle (frequency) - the beam angle for each frequency

> 

Reviewer: ```suggestion
Note that the key data variables you will be working with are `Sv` and `range`. All other variables are included in this dataset so that the parameters used in the calibration (from raw to Sv) are recorded.
```

> Using the above Dataset we can visualize the reflection of the echo (often called the echogram) by plotting Sv: 

Reviewer: ```suggestion
Using the above Dataset we can visualize the strength of the echoes (often called the echogram) by plotting Sv: 
```

![echogram example](./echogram_example.png)

By compiling several of these echograms and processing the data further, one can visualize the data over several hours.
This can yield visualizations such as the below image, which shows the impact of the eclipse-driven reduction in 
sunlight on marine zooplankton: 

![](./bokeh_plot.png) 

### Additional Resources

Some useful resources for getting started with the proposed visualization tools: 

* Getting started with [HoloViz](https://nbviewer.org/github/philippjfr/pydata-2021/blob/master/PyData_2021.ipynb)

* Useful resources and example dashboards in [Panel](https://awesome-panel.org/)

### Initial Steps

1. Read the example files provided in TBD using xarray

2. Construct a widget that displays the Sv variable with ping_time as the x coordinate and range_bin as the y 
coordinate
 
3. Improve the widget by allowing the user to change the frequency and the colormap

> 4. Explore the desired types of visualization in gsoc2022_wanted

Reviewer: ```suggestion
4. Explore the desired types of visualization -- these are issues labeled with `gsoc 2022 wanted`
```

## Getting Started

As this project is an ongoing development of echopype, there are a couple of preliminary items you should become familiar
with before diving in. We outline these items below as well as provide some additional helpful resources. 

### Storage format

In echopype, we utilize two methods for storing results: 

* netcdf files - the current defacto file for working with array-oriented scientific data from climate research.
Although it is not necessary to understand the netCDF library in its entirety, Unidata does provide a well documented 
[netcdf](https://unidata.github.io/netcdf4-python/) python interface. This documentation describes useful aspects of 
how netcdf defines common terms such as groups, dimensions, variables, and attributes.  

* [Zarr](https://zarr.readthedocs.io/en/stable/) - a format for the storage of chunked, compressed, N-dimensional 
arrays. Zarr has similar characteristics to netcdf, but has the added benefit of being able to be both read and wrote to

> concurrently from multiple threads or processes. Zarr provides a great overview of its [storage specifications](https://zarr.readthedocs.io/en/stable/spec/v2.html#hierarchies)

Reviewer 2: I don't think it's necessary to go into concurrency/threads/processes here. Just the chunking part and some comments about cloud-native will suffice.

that may be useful to read.    

> 

Reviewer 2: Just like Valentina said. Zarr is a cloud-native data format. Which means that it's made for storing things in the cloud.

* range_bin - sample index along a range

The data variables of the Dataset are as follows, where items in parenthesis are the dimensions of the data variables: 

* Sv (frequency, ping_time, range_bin) - volume backscattering strength measured from the echo

* range (frequency, ping_time, range_bin) - the measured range of an echo in meters 

* temperature - the temperature measurement of the water collected by the echosounder, with unit degree Celsius

* salinity - the salinity measurement of the water collected by the echosounder, with unit part per thousand (PSU)

* pressure - the pressure measurement of the water collected by the echosounder, with unit dbars

* sound_speed (frequency, ping_time) - sound speed (in units m/s) for the provided temperature, salinity, and pressure 

* sound_absorption (frequency, ping_time) - sea water absorption (in units dB/m) for each frequency and ping time, this 
value is based on the temperature, salinity, pressure, and sound_speed

* sa_correction (frequency) - the sa correction for each frequency

* gain_correction - (frequency) - the gain correction for each frequency

* equivalent_beam_angle (frequency) - the beam angle for each frequency

> 

Author: I am not sure which data we plan to use. I believe that we still need to put together the data, see issue #3. I am currently using the newest stable version of Echopype and data from the Hake survey. I am not sure how this output changes with respect to the data. Maybe @leewujung can provide some insight here.

> 

Reviewer: @valentina-s: these variables were added in one of the v0.5.x releases so that the parameters that were used to calibrate the raw data into Sv are recorded. 

I think we should make sure all the datasets we share in this repo have the same structure/variables to avoid confusion. How many files are used in the echogram notebook you mentioned in #3 ?  It'll probably be a good idea that we re-convert them.

> 

Reviewer: @b-reyes : I think here it would help to use the inline backtick (like Sv --> `Sv`) so that contributors know that there are variables

### Additional Resources

Some useful resources for getting started with the proposed visualization tools: 

* Getting started with [HoloViz](https://nbviewer.org/github/philippjfr/pydata-2021/blob/master/PyData_2021.ipynb)

* Useful resources and example dashboards in [Panel](https://awesome-panel.org/)

> 

Reviewer 2: We'll have to work on making these notebooks next I think.
